utils/logalloc.cc

```diff
@@ -360,9 +364,76 @@ static thread_local tracker& tracker_instance = get_tracker_instance();
 
 using clock = std::chrono::steady_clock;
 
+class background_reclaimer {
+    scheduling_group _sg;
+    noncopyable_function<void (size_t target)> _reclaim;
+    timer<lowres_clock> _adjust_shares_timer;
+    // If engaged, main loop is not running, set_value() to wake it.
+    promise<>* _main_loop_wait = nullptr;
+    future<> _done;
+    bool _stopping = false;
+    static constexpr size_t free_memory_threshold = 60'000'000;
+private:
+    bool have_work() const {
+#ifndef SEASTAR_DEFAULT_ALLOCATOR
+        return memory::stats().free_memory() < free_memory_threshold;
+#else
+        return false;
+#endif
+    }
+    void main_loop_wake() {
+        llogger.debug("background_reclaimer::main_loop_wake: waking {}", bool(_main_loop_wait));
+        if (_main_loop_wait) {
+            _main_loop_wait->set_value();
+            _main_loop_wait = nullptr;
+        }
+    }
+    future<> main_loop() {
+        llogger.debug("background_reclaimer::main_loop: entry");
+        while (true) {
+            while (!_stopping && !have_work()) {
+                promise<> wait;
+                _main_loop_wait = &wait;
+                llogger.trace("background_reclaimer::main_loop: sleep");
+                co_await wait.get_future();
+                llogger.trace("background_reclaimer::main_loop: awakened");
+                _main_loop_wait = nullptr;
+            }
+            if (_stopping) {
+                break;
+            }
+            _reclaim(free_memory_threshold - memory::stats().free_memory());
+            co_await make_ready_future<>();
+        }
+        llogger.debug("background_reclaimer::main_loop: exit");
+    }
+    void adjust_shares() {
+        if (_main_loop_wait && have_work()) {
+            _sg.set_shares(1 + (1000 * (free_memory_threshold - memory::stats().free_memory())) / free_memory_threshold);
+            main_loop_wake();
+        }
+    }
+public:
+    explicit background_reclaimer(scheduling_group sg, noncopyable_function<void (size_t target)> reclaim)
+            : _sg(sg)
+            , _reclaim(std::move(reclaim))
+            , _adjust_shares_timer(_sg, [this] { adjust_shares(); })
+            , _done(with_scheduling_group(_sg, [this] { return main_loop(); })) {
+        if (sg != default_scheduling_group()) {
+            _adjust_shares_timer.arm_periodic(50ms);
```

Comment:
shouldn't we increase the frequency if we're close to threshold? i am worried that 50ms may be too late if we're close to the threshold. example: let's say clock triggers now, we have 65MB of free memory, and in the next 50ms, we can easily allocate that memory before background reclaiming has any chance to run. perhaps we should make <100mb the red zone threshold, at which we'll check for reclaiming work more often, using perhaps a 10/20ms clock interval instead.

Reply:
We'll always be close to the threshold, since lsa is continuously allocating. What I want is to have a notifier in seastar to call us back on a configurable threshold. Seastar checks its own threshold anyway, so we can get that check for free.